packages/pwa-kit-extension-sdk/src/react/classes/ApplicationExtension.ts

```diff
@@ -73,4 +83,106 @@ export class ApplicationExtension<
     public beforeRouteMatch(routes: RouteProps[]): RouteProps[] {
         return routes
     }
+
+    /**
+     * Called on the server to serialize the extension data that will be sent to the client.
+     * 
+     * @returns SerializedExtension - The serialized extension data.
+     * @throws Error if the routes have not been loaded.
+     */
+    public serialize(): SerializedExtension {
+        if (this._cachedRoutes === null) {
+            throw new Error('Routes have not been loaded. Call getRoutes() before serializing')
+        }
+        const serializedRoutes = this._cachedRoutes.map((route) => {
+            if (!route.component?.displayName) {
+                throw new Error(`Component for route with path "${route.path}" is missing a displayName in ${this.getName()} extension`)
+            }
+            return {
+                path: route.path,
+                componentName: route.component?.displayName,
+            }
+        })
+        return {
+            routes: serializedRoutes
+        }
+    }
+
+    /**
+     * Returns a map of component names to components that are used to deserialize the extension data.
+     * 
+     * @protected
+     * @returns ComponentMap - The map of component names to components.
+     */
+    protected getComponentMap?(): ComponentMap
+
+    /**
+     * Called on the client to deserialize the extension data that was serialized on the server.
+     * 
+     * @param serializedExtension - The serialized extension data.
+     * @returns DeserializedExtension - The deserialized extension data.
+     * @throws Error if getComponentMap() is not defined.
+     * @throws Error if the deserialized component cannot be found in the component map.
+     */
+    private deserialize(serializedExtension: SerializedExtension): DeserializedExtension {
+        if (!this.getComponentMap) {
+            throw new Error(`${this.getName()}.getRoutes() is async but does not define getComponentMap()`);
+        }
```

Comment:
rather than enforcing a `getComponentMap` impl in the constructor I moved it to the deserialize method (which is called in the constructor) to avoid the typescript error on line 132 as getComponentMap can now be undefined